service/controller/v21/adapter/guest_auto_scaling_group.go

```diff
@@ -22,9 +23,19 @@ type GuestAutoScalingGroupAdapter struct {
 }
 
 func (a *GuestAutoScalingGroupAdapter) Adapt(cfg Config) error {
-	workers := key.WorkerCount(cfg.CustomObject)
-	if workers <= 0 {
-		return microerror.Maskf(invalidConfigError, "at least 1 worker required, found %d", workers)
+	maxWorkers := key.ScalingMax(cfg.CustomObject)
+	minWorkers := key.ScalingMin(cfg.CustomObject)
+
+	if minWorkers <= 0 {
+		return microerror.Maskf(invalidConfigError, "at least 1 worker required, found %d", minWorkers)
+	}
+
+	if maxWorkers < minWorkers {
+		return microerror.Maskf(invalidConfigError, "maximum number of workers (%d) is smaller than minimum number of workers (%d)", maxWorkers, minWorkers)
+	}
+
+	if maxWorkers == minWorkers {
+		maxWorkers++
```

Comment:
Why are we doing this here?

Reply:
Because in previous version (little bit down here below) it used to be like this:
```go
	a.ASGMaxSize = workers + 1
	a.ASGMinSize = workers
```

My guess for why this is like this is that this gives room for cluster movement when rolling nodes for any reason (although upgrade is handled still little bit differently), but I might be wrong. Nevertheless, that's how it used to be.

Reply:
Might be legacy from before we started using cloudformations.

Reply:
No, it's actually needed and we discussed about this. I also added comment here, but this source code snippet in discussion view is old.

Reply:
The upgrade first starts new instances. Once they are up and running we start with the draining. On bigger clusters we definitely start bigger batches than a single node. 

I just started a test cluster with 12 nodes on AWS. Max was set to 13. The rolling update of the ASG started 4 new workers.

Are you really sure Max has any effect on rolling updates?

Reply:
Settled with max == min for now. After discussions this should be correct nowadays with lifecycle hooks etc.. Testing will prove this a bit later.